### ISSUE TITLE
Donald Trump Tweet Bot

### ISSUE THREAD

**Issue by jvivian** [jvivian/Children-of-Skynet#2]
- [x] (John) Create directory for project
- [x] (John/Sam) Collect tweets -- file, one line per tweet?
- [x] (John) Sanitize/preprocess tweets 
- [x] (John) Create notebook of some basic data exploration
- [x] (John) Write RNN training module
- [x] (John) Write text generation module
- [x] (John) Train models on GPU cluster
- [ ] (Jake) Model refinement / model choice?
- [ ] (Sam) Create account for twitter bot 
- [ ] (Sam) Code for twitter bot to grab generated tweet and post

A few things to work out or optional challenges:

- How to automate the twitter bot to post everyday (scheduled job? Someone have a server we can use?)
- Incorporating new tweets — train model on new tweets as they show up.
- Context posts — We could combine the tweet generator with a heuristic to check for relevant topics so it posts things that appear "current event". 

Let me know what ya think — we can of course change who works on what, etc.

**Comment by jvivian** [comment]
@nsjake — Tweets are in the repo, you can start working on your own notebook if you want. My work in progress is located in lstm.ipynb

@avacabro @AaronJCooke — I made a notebook called tweet-analysis. I tried to comment _a lot_ to explain how I use pandas to answer questions I have about the data. Let me know if anything is unclear.

**Comment by jvivian** [comment]
Some stats

- ~32,000 tweets
- ~3.3 million characters
- ~1.1 million training examples using a length of 40 and a stride of 3.
- Training on a non-GPU machine takes about 1 hour per epoch. 20 epochs are recommended as a good starting point.

**Comment by jvivian** [comment]
Hey guys — so I've uploaded several trained models to the repo under **twit_bot/models/sanitized-tweets**. You can play around with them by doing:

`model = keras.models.load_model()` and providing a path to the hdf5 model file in the subdirectories. The subdir names represent the parameter settings the model was trained with. 

I added a new CLI that you can access via `generate-text --help` that will accept a model and the original input text to generate text for you.  The key is playing with the `--temp` parameter which introduces multinomial sampling of the classification vector to introduce entropy during the generation process. This addresses the biggest issue these generative models have which is getting stuck in infinite loops.  I added an explanation of how the sampling works in the [docstring of the sampling function.](https://github.com/jvivian/Children-of-Skynet/blob/master/twit_bot/generate_text.py#L9-L18)

**Comment by jvivian** [comment]
At this point, the backend model is getting close to a "useable" form.  We should start thinking about logistics for:

- Running the twitter bot
- How we want to seed the model generation
- Other heuristics for determining final tweet output

@nsjake  / @avacabro  — Either of you interested or have the bandwidth to work on this part?
